### PR TITLE
CORE-7769 utxo finality

### PR DESCRIPTION
**Pull request by vlajos**
* Mainly porting Consensual code:
  * `UtxoFinalityFlow`, `UtxoReceiveFinalityFlow`
   * ledger-utxo-demo-app
   * Utxo Ledger smoke test
* Other small fixes (`@Suspendable`s, etc.)

Missing pieces of the Finality flow:
* CORE-8271 NotaryLookup looks broken. Until that I just added a member as notary to be able to proceed.
* ~~CORE-7270 Utxo Signatories are not handled correctly. So the related test is disabled for now.~~
* E2E tests time out in jenkins. Not too sure if it is caused by this change. But the extra test may need extra time too.
* Update flow based on the diagram:
  * Add the remaining persistence points.
  * Add notarisation step.
  * Add missing signature distribution steps.
  * Add latest database/vault update steps.
  * Confirm that verifications are in place.
  * Update tests accordingly.
* Update diagram/code:
  * Expand error-cases, implement, test them.

### CODE REVIEW

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/UtxoFinalityFlow.kt [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            session to member
+        }.associate { (session, memberInfo) ->
+            session to memberInfo.ledgerKeys.ifEmpty {
+                throw CordaRuntimeException(
+                    "A session with ${session.counterparty} exists but the member does not have any active ledger keys"
+                )
+            }
+        }
+
+        // Should this also be a [CordaRuntimeException]? Or make the others [IllegalArgumentException]s?
```
Comment (lankydan) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/UtxoFinalityFlow.kt]: I'm going to make a tech debt for this.

Comment (lankydan) [reply components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/UtxoFinalityFlow.kt]: https://r3-cev.atlassian.net/browse/CORE-8655

File: libs/ledger/ledger-utxo-data/src/main/kotlin/net/corda/ledger/utxo/data/transaction/UtxoTransactionVerification.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * Centralized place for transactions verifications.
+ * They are supposed to be called
+ *  - at the transaction's creation time from [UtxoTransactionBuilder] or [UtxoSignedTransactionFactory]
+ *  - or later on the receiving side of the Finality flow. ([UtxoReceiveFinalityFlow])
+ *
+ */
+
+class UtxoTransactionVerification {
```
Comment (lankydan) [inline libs/ledger/ledger-utxo-data/src/main/kotlin/net/corda/ledger/utxo/data/transaction/UtxoTransactionVerification.kt]: This doesn't feel like the right place for this. Should this be in the flow module?

Comment (vlajos) [reply libs/ledger/ledger-utxo-data/src/main/kotlin/net/corda/ledger/utxo/data/transaction/UtxoTransactionVerification.kt]: I have moved it there.

**Review by lankydan — APPROVED** [review]

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/UtxoFinalityFlow.kt [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+        // Check if the sessions' counterparties are all available and have keys.
+        val sessionPublicKeys = sessions.map { session ->
+            val member = memberLookup.lookup(session.counterparty)
+                ?: throw CordaRuntimeException(
+                    "A session with ${session.counterparty} exists but the member no longer exists in the membership group"
+                )
+            session to member
+        }.associate { (session, memberInfo) ->
+            session to memberInfo.ledgerKeys.ifEmpty {
```
Comment (mnesbit) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/UtxoFinalityFlow.kt:58]: We probably shouldn't do this check as it won't apply when we stop forcing use of ledger keys only. Instead we have to check all signatures are present after collecting signatures and before going to notary.

File: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/UtxoFinalityFlow.kt [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+                throw CordaRuntimeException(
+                    "A session with ${session.counterparty} exists but the member does not have any active ledger keys"
+                )
+            }
+        }
+
+        // TODO [CORE-8655] Should this also be a [CordaRuntimeException]? Or make the others [IllegalArgumentException]s?
+        val missingSignatories = signedTransaction.getMissingSignatories()
+        // Check if all missing signing keys are covered by the sessions.
+        require(sessionPublicKeys.values.flatten().containsAll(missingSignatories)) {
```
Comment (mnesbit) [inline components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/flows/finality/UtxoFinalityFlow.kt:68]: Again this check isn't helpful and we should probably remove. If the caller of the flow doesn't pass in correct sessions list, or those sessions don't sign it isn't our fault.

**Review by ronanbrowne — APPROVED** [review]
LGTM (Build)

**Pull request merged**